standards/application/relay-reliability.md

```diff
@@ -0,0 +1,150 @@
+---
+title: RELIABILITY-RELAY
+name: Reliability for Relay Protocol
+category: Standards Track
+tags: [reliability, application]
+editor: Kaichao Sun <kaichao@status.im>
+contributors:
+  - Richard Ramos <richard@status.im>
+---
+
+## Abstract
+
+[Relay Protocol](https://github.com/vacp2p/rfc-index/blob/main/waku/standards/core/11/relay.md) in Waku is efficient for routing messages, but there's no guarantee that a message will reach to its destination. For example, the receiver in a chat application may miss messages when network issue happens at either sender or receiver side. In general, a message in Waku network includes 3 status:
+
+- **outgoing**, the message is posted by its creator but no confirmations from other nodes yet
+- **sent**, the message is received by any other node in the network
+- **delivered**, the message is acknowledged by the receiver
+
+Application like Status already uses [MVDS](https://github.com/vacp2p/rfc-index/blob/main/vac/2/mvds.md) for e2e acknowledgement in direct messages and group chat. There is an ongoing [discussion](https://forum.vac.dev/t/end-to-end-reliability-for-scalable-distributed-logs/293) about a more general and bandwidth efficient solution for e2e reliablity.
+
+Before we have a complete design for e2e reliability, we need to compose existing protocols to increase the reliability of the relay protocol. This document proposes a few options for such composition.
+
+## Motivation
+
+The [Store protocol](https://github.com/waku-org/specs/blob/master/standards/core/store.md) provides a way for nodes in the network to query the existence of messages or fetch specific messages based on the search criteria.
+
+**Search criteria with message hash**
+
+For the nodes that may have connection issues to **publish** messages via relay network, this search criteria can be used to check whether a message is populated in the network or not. The message exists in store node can be marked from `outgoing` to `sent` by application. If the message is not found in the store node, the application can resend the message.
+
+**Search criteria with topics and time range**
+
+For the nodes that may have connection issues to **receive** messages via relay network, this search criteria can be used to fetch missing messages from store nodes when network resumes. 
+
+By leveraging the store node to provide such query services, the applications are able to mitigate the reliability issue of relay protocol. But this approach also introduces new challenges like centralization, privacy, and scalability. It should be viewed as a temporary solution and deprecated when e2e reliability solution is ready.
+
+## Implementation Suggestions 
```

Comment:
An Implementation Suggestion can be added towards the end of the spec, but I'd describe the mechanism here in a simple heuristic. Also, since we already do periodic store queries to find missing received messages, I don't see the need for the "queries with message hash" for outgoing messages. Why not simply verify that your outgoing message hashes are included in the periodic query results? This way we can have a consolidated mechanism that deals with both outgoing/publish and incoming/received message reliability.

In fact, I think the heuristic will be simpler to describe than an implementation suggestion. Very roughly:

1. create outgoing buffer for all "outgoing" messages
2. create incoming buffer for all recently received message hashes
3. periodic store query based on time range, returning only message hashes
4. check outgoing buffer: message hashes included in response from (3) are "sent", others are resent
5. check incoming buffer: message hashes included in response from (3) that's not in incoming buffer are missing. Request these from the store.

There will of course be more details, but I think the main focus of this spec should be a simple, consolidated mechanism close to the above. WDYT?

Reply:
> to do such refactor in status-go is not a trivial work

Valid concern, though I would focus on describing the concept as simply (and puristically) as possible in the spec even if our initial implementation(s) may not fully implement the spec. In other words, I wouldn't write a spec based on the limitations of the implementation, but extract the concepts in such a way that we can, incrementally, bring the implementation to fully implement the spec even if for now we start with what is trivial to add.

> the requirement is different

This is indeed a very good point. In a heuristic like above, this would add an optional mechanism for more regular (and even more cheap) outgoing message checks.